File: huffman_Encoder_Decoder.hs
import System.IO

-- Custom frequency count function
countElem :: (Eq a) => [a] -> a -> Int
countElem [] _ = 0
countElem (x:xs) e = if x == e then 1 + countElem xs e else countElem xs e

-- Filter out elements equal to the given value
filterElem :: (Eq a) => [a] -> a -> [a]
filterElem xs e = filter (/= e) xs

-- Generate frequency list
freqList :: Eq a => [a] -> [(a, Int)]
freqList [] = []
freqList (x:xs) = (x, countElem (x:xs) x) : freqList (filterElem xs x)

-- Tree data structure
data HuffmanTree = HuffmanLeaf Char | HuffmanNode Int HuffmanTree HuffmanTree deriving Show

-- Generate Huffman Tree
buildHuffmanTree :: [(Char, Int)] -> HuffmanTree
buildHuffmanTree xs = constructTree (sortNodes (map (\(x, y) -> (y, HuffmanLeaf x)) xs))

-- Sorting helper functions
sortNodes :: [(Int, HuffmanTree)] -> [(Int, HuffmanTree)]
sortNodes = foldr insertNode []

insertNode :: (Int, HuffmanTree) -> [(Int, HuffmanTree)] -> [(Int, HuffmanTree)]
insertNode x [] = [x]
insertNode x (y:ys) = if fst x <= fst y then x : y : ys else y : insertNode x ys

-- Combining trees in building Huffman Tree
combineTrees :: (Int, HuffmanTree) -> (Int, HuffmanTree) -> (Int, HuffmanTree)
combineTrees (p1, t1) (p2, t2) = (p1 + p2, HuffmanNode (p1 + p2) t1 t2)

-- Build Huffman Tree recursively
constructTree :: [(Int, HuffmanTree)] -> HuffmanTree
constructTree [(_, t)] = t
constructTree (a:b:xs) = constructTree (insertNode (combineTrees a b) xs)

-- Generating Huffman codes
generateHuffmanCodes :: HuffmanTree -> [(Char, String)]
generateHuffmanCodes tree = getPath tree ""

getPath :: HuffmanTree -> String -> [(Char, String)]
getPath (HuffmanLeaf s) path = [(s, path)]
getPath (HuffmanNode _ l r) path = getPath l (path ++ "0") ++ getPath r (path ++ "1")

-- Encoding string using Huffman codes
huffmanEncode :: String -> String
huffmanEncode xs = concatMap (\x -> encodeChar (generateHuffmanCodes (buildHuffmanTree (freqList xs))) x) xs

-- Decoding string using Huffman codes
huffmanDecode :: String -> String
huffmanDecode xs = decode xs (buildHuffmanTree (freqList xs))
  where
    decode [] _ = ""
    decode xs tree = case decodeChar xs tree of
                      Just (char, rest) -> char : decode rest tree
                      Nothing           -> error "Invalid Huffman encoding"

    decodeChar xs (HuffmanLeaf s) = Just (s, xs)
    decodeChar ('0':xs) (HuffmanNode _ l _) = decodeChar xs l
    decodeChar ('1':xs) (HuffmanNode _ _ r) = decodeChar xs r
    decodeChar _ _ = Nothing

-- Modified `ord` behavior to mimic 'Data.Char.ord' for '0'-'9'
-- Directly subtracting 48 assuming input will be '0'-'9'
charToDigit :: Char -> Int
charToDigit x = fromEnum x - 48

-- Example Main to demonstrate functionality
main :: IO ()
main = do
  putStrLn "Enter the name of the text file (with extension):"
  fileName <- getLine
  contents <- readFile fileName
  putStrLn "Original Text:"
  putStrLn contents
  putStrLn "Huffman Encoded:"
  let encoded = huffmanEncode contents
  putStrLn encoded
  putStrLn "Huffman Decoded:"
  let decoded = huffmanDecode encoded
  putStrLn decoded

File: huffman_Encoder_Decoder.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

// Custom frequency count function
template <typename T>
int countElem(const vector<T>& xs, T e) {
    int count = 0;
    for (const auto& x : xs) {
        if (x == e) {
            count++;
        }
    }
    return count;
}

// Filter out elements equal to the given value
template <typename T>
vector<T> filterElem(const vector<T>& xs, T e) {
    vector<T> filtered;
    for (const auto& x : xs) {
        if (x != e) {
            filtered.push_back(x);
        }
    }
    return filtered;
}

// Generate frequency list
template <typename T>
vector<pair<T, int>> freqList(const vector<T>& xs) {
    vector<pair<T, int>> freqs;
    for (const auto& x : xs) {
        int count = countElem(xs, x);
        auto it = find_if(freqs.begin(), freqs.end(), [x](const pair<T, int>& p) {
            return p.first == x;
        });
        if (it == freqs.end()) {
            freqs.push_back(make_pair(x, count));
        }
    }
    return freqs;
}

// Tree data structure
struct HuffmanTree {
    char character;
    int frequency;
    HuffmanTree* left;
    HuffmanTree* right;
    HuffmanTree(char c, int freq) : character(c), frequency(freq), left(nullptr), right(nullptr) {}
};

// Sorting helper function for nodes
bool sortByFrequency(const pair<int, HuffmanTree*>& a, const pair<int, HuffmanTree*>& b) {
    return a.first < b.first;
}

// Generate Huffman Tree
HuffmanTree* buildHuffmanTree(const vector<pair<char, int>>& freqs) {
    vector<pair<int, HuffmanTree*>> nodes;
    for (const auto& p : freqs) {
        nodes.push_back(make_pair(p.second, new HuffmanTree(p.first, p.second)));
    }

    while (nodes.size() > 1) {
        sort(nodes.begin(), nodes.end(), sortByFrequency);
        auto first = nodes[0];
        auto second = nodes[1];
        nodes.erase(nodes.begin(), nodes.begin() + 2);

        HuffmanTree* combined = new HuffmanTree('\0', first.first + second.first);
        combined->left = first.second;
        combined->right = second.second;

        nodes.push_back(make_pair(first.first + second.first, combined));
    }

    return nodes[0].second;
}

// Generating Huffman codes
void generateHuffmanCodes(HuffmanTree* tree, map<char, string>& codes, string path) {
    if (tree->left == nullptr && tree->right == nullptr) {
        codes[tree->character] = path;
    } else {
        generateHuffmanCodes(tree->left, codes, path + "0");
        generateHuffmanCodes(tree->right, codes, path + "1");
    }
}

// Encoding string using Huffman codes
string huffmanEncode(const string& xs, const map<char, string>& codes) {
    string encoded;
    for (char c : xs) {
        encoded += codes.at(c);
    }
    return encoded;
}

// Decoding string using Huffman codes
string huffmanDecode(const string& xs, HuffmanTree* tree) {
    string decoded;
    HuffmanTree* current = tree;
    for (char bit : xs) {
        if (bit == '0') {
            current = current->left;
        } else if (bit == '1') {
            current = current->right;
        }

        if (current->left == nullptr && current->right == nullptr) {
            decoded += current->character;
            current = tree;
        }
    }
    return decoded;
}

int main() {
    cout << "Enter the name of the text file (with extension):" << endl;
    string fileName;
    cin >> fileName;

    ifstream file(fileName);
    if (!file.is_open()) {
        cerr << "Error opening file." << endl;
        return 1;
    }

    string contents((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
    file.close();

    cout << "Original Text:" << endl;
    cout << contents << endl;

    auto freqs = freqList(contents);

    HuffmanTree* tree = buildHuffmanTree(freqs);

    map<char, string> codes;
    generateHuffmanCodes(tree, codes, "");

    cout << "Huffman Encoded:" << endl;
    string encoded = huffmanEncode(contents, codes);
    cout << encoded << endl;

    cout << "Huffman Decoded:" << endl;
    string decoded = huffmanDecode(encoded, tree);
    cout << decoded << endl;

    return 0;
}
